keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,532 @@
+  controller) create Endpoints objects.
+
+- Update the e2e test suite to ensure that all tests that care about
+  endpoints only as a generic concept are using EndpointSlices rather
+  than Endpoints. (For example, the utility function
+  `framework.WaitForServiceEndpointsNum` should check that
+  EndpointSlices exist, not that Endpoints exist).
+
+- Adjust the conformance test suite to not require the Endpoints and
+  EndpointSlice Mirroring controllers to be running:
+
+    - Fix tests such as `Service endpoints latency should not be very
+      high` to check the latency of EndpointSlices rather than
+      Endpoints, since that is what is actually important in
+      real-world usage.
+
+    - Demote tests such as `EndpointSliceMirroring should mirror a
+      custom Endpoints resource through create update and delete` to
+      non-conformance.
+
+    - Split up tests such as `EndpointSlice should create Endpoints
+      and EndpointSlices for Pods matching a Service` into a
+      conformance test for EndpointSlice and a non-conformance test
+      for Endpoints.
+
+- Update the e2e test suite so that all remaining tests that depend on
+  the outputs of the Endpoints or EndpointSlice Mirroring controllers
+  are feature-tagged so they can be skipped in clusters that do not
+  run those controllers, and set up a periodic e2e job running in such
+  a configuration.
+
+- Explicitly document that disabling `endpoints-controller` and/or
+  `endpointslice-mirroring-controller` via kube-controller-manager's
+  `--controllers` flag is a supported and conforming configuration.
+
+    - Figure out what to do with stale Endpoints in such a cluster.
+
+```
+<<[UNRESOLVED kubernetes.default ]>>
+
+- MAYBE change kube-apiserver to optionally not generate Endpoints for
+  kubernetes.default, though this would require adding a
+  kube-apiserver configuration option, and the benefit from removing
+  just that 1 Endpoints object is small. Perhaps instead it could just
+  add an annotation to the object pointing out the fact that Endpoints
+  is deprecated.
+
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED disable-by-default ]>>
+
+- MAYBE eventually move `endpoints-controller` and
+  `endpointslice-mirroring-controller` out of the "on-by-default"
+  controller list so that they become opt-in rather than opt-out.
+
+- MAYBE eventually switch our e2e test jobs to disable those
+  controllers by default, and only run a single periodic/optional job
+  to run the Endpoints-specific tests.
+
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Deleting or modifying the `v1.Endpoints` API.
+
+## Proposal
+
+Overall, this KEP is mostly just about documentation and tests; it is
+already possible to disable the Endpoints and EndpointSlice Mirroring
+controllers via kube-controller-manager's `--controllers` option, and
+we believe that this will have no ill effects in a vanilla Kubernetes
+cluster (though, currently, it will cause the e2e tests to fail).
+
```

Comment:
If a controller is (still) making Endpoints objects, do we want to surface that somehow once the API is deprecated?

eg: `.status` on the associated Service?

Reply:
Yes, maybe? But...

In my first draft I suggested that it should add an annotation to each Endpoints:

```
metadata:
  annotations:
    endpoints.kubernetes.io/deprecated: >
      The Endpoints API is deprecated, and all users should use the
      EndpointSlice API instead. Endpoints objects are not guaranteed
      to be generated in all clusters.
```

but then I got to the PRR question about "Will enabling / using this feature result in increasing size or count of the existing API objects?" and was like, oh, yeah, probably we shouldn't be adding an annotation to every Endpoints when part of the motivation is to stop wasting space in etcd on objects nobody looks at...

(Yay PRR questions!)

(Then I thought about possibly hacking the annotations in on Get and stripping them back out on Update but that seemed potentially fragile...)

Reply:
I'm OK with `kubernetes.io/deprecated` as an annotation. Even if it makes the objects larger.